src/v/cloud_topics/data_plane_api.h

```diff
@@ -37,11 +56,17 @@ class data_plane_api {
     virtual ss::future<> start() = 0;
     virtual ss::future<> stop() = 0;
 
-    /// Write data batches and get back placeholder batches
-    virtual ss::future<result<chunked_vector<extent_meta>>> write_and_debounce(
+    // Reserve the space needed for this write.
+    virtual ss::future<std::expected<write_reservation, std::error_code>>
+    reserve_write(chunked_vector<model::record_batch> batches) = 0;
```

Comment:
This could return units (wrapped in `write_reservation`) and then `execute_write` could accept both batches and the reservation.  Also, this adds additional allocation to the write path.

Reply:
I'm not too worried about a single allocation, we're buffering everything in memory, and moving a future around can add an allocation.

> This could return units (wrapped in write_reservation) and then execute_write could accept both batches and the reservation.

But reservation needs to serialize the data, what do you expect to do there?

Reply:
Yeah I renamed to staged_*

Reply:
How the shutdown is handled? If the queue is deep do we have to wait until every trekt is released sequentially or the whole queue will be released at once?